File: android/app/src/main/kotlin/com/poracode/app/transport/RemoteMutationClassification.kt
package com.poracode.app.transport

import com.poracode.app.model.RemoteClientException

/**
 * Single rule for mutation-outcome ambiguity across every remote domain.
 *
 * A mutation is ambiguous — [requestMayHaveCommitted] — when the request could have
 * reached the server and been applied before the failure was observed: HTTP >= 500,
 * no response at all (status <= 0), a client-side transport/decode classification
 * (network, timeout, invalid response, oversize response), or the host's explicit
 * uncertain-receipt code. 4xx rejections, scope and validation failures are
 * definite, and reads are always definite regardless of status.
 */
object RemoteMutationClassification {
    /** Client- and host-assigned codes that mean the outcome of a sent mutation is unknown. */
    val ambiguousCodes: Set<String> = setOf(
        "network",
        "timeout",
        "invalid_response",
        "response_too_large",
        /**
         * The host's typed 409 for an idempotent command whose external effect may
         * already have happened (interrupted dispatch, unresolved receipt). It is
         * a truthful ambiguity, never a definite rejection, so the UI must keep
         * `requestMayHaveCommitted` and reconcile instead of offering a plain retry.
         */
        "command_outcome_uncertain",
    )

    fun isAmbiguousOutcome(status: Int, code: String): Boolean =
        status <= 0 || status >= 500 || code in ambiguousCodes

    fun requestMayHaveCommitted(status: Int, code: String, mutation: Boolean): Boolean =
        mutation && isAmbiguousOutcome(status, code)

    fun requestMayHaveCommitted(error: RemoteClientException, mutation: Boolean): Boolean =
        requestMayHaveCommitted(error.status, error.code, mutation)
}

File: ios/App/App/Transport/RemoteMutationClassification.swift
import Foundation

/// Central classification of remote mutation failures.
///
/// A mutation is submitted exactly once. When the outcome cannot be established
/// after submission, the request may have been committed on the host: callers must
/// never replay it and must reconcile with exactly one authoritative read instead.
enum RemoteMutationClassification: Sendable, Equatable {
  /// The host's explicit uncertain-outcome code for an idempotent command whose
  /// external effect may already have happened (interrupted dispatch, or a
  /// receipt that is still unresolved). It is a truthful ambiguity, never a
  /// definite rejection.
  static let uncertainOutcomeCode = "command_outcome_uncertain"

  /// The request could have reached the server and been committed: HTTP >= 500,
  /// status 0 / network / timeout failures, the host's explicit uncertain
  /// outcome, or a response that cannot establish the outcome after the
  /// mutation was sent.
  case requestMayHaveCommitted

  /// The outcome is definite: the mutation was not committed (4xx, scope and
  /// validation rejections, pre-send failures). Reads always classify this way.
  case definiteFailure

  /// Classify a post-send HTTP failure for a mutation. The host error code is
  /// carried alongside the status, so an explicit `command_outcome_uncertain`
  /// 409 stays ambiguous even though ordinary 409s are definite.
  static func classify(statusCode: Int, code: String? = nil) -> Self {
    if statusCode == 0 { return .requestMayHaveCommitted }
    if statusCode >= 500 { return .requestMayHaveCommitted }
    if code == "network" || code == "timeout" { return .requestMayHaveCommitted }
    if code == uncertainOutcomeCode { return .requestMayHaveCommitted }
    return .definiteFailure
  }

  /// Reads carry no commit semantics: every read failure is definite, including 5xx.
  static func classifyRead(statusCode: Int, code: String? = nil) -> Self {
    _ = statusCode
    _ = code
    return .definiteFailure
  }

  /// A transport drop after submission (no usable response) cannot establish the outcome.
  static var transportDrop: Self { .requestMayHaveCommitted }

  /// True when a post-send mutation failure is ambiguous and the caller must reconcile.
  static func isAmbiguous(statusCode: Int, code: String? = nil) -> Bool {
    classify(statusCode: statusCode, code: code) == .requestMayHaveCommitted
  }
}
